File: src/hypeedge/backtest/engine.py
"""Backtest engine — orchestrates strategy, broker, and data feed for simulation.

Replaces the Phase 1 skeleton with a fully functional event-driven backtest
that reuses the same StrategyBase.on_event() code path as live trading.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any

import structlog

from hypeedge.backtest.broker import (
    FeeConfig,
    SimulatedBroker,
    SlippageConfig,
    SlippageMode,
)
from hypeedge.backtest.data_feed import DataFeed
from hypeedge.backtest.metrics import MetricsCalculator, PerformanceMetrics
from hypeedge.core.enums import OrderStatus, Side
from hypeedge.core.events import (
    EVENT_CANDLE_UPDATE,
    EVENT_ORDER_CANCELLED,
    EVENT_ORDER_FILLED,
    EVENT_ORDER_SUBMITTED,
    Event,
    EventBus,
)
from hypeedge.core.models import Candle, Fill, FundingRate, Order, OrderIntent, Position
from hypeedge.core.types import Cloid, Price, Size, Timestamp, Usd

logger = structlog.get_logger(__name__)

# Type alias for the strategy factory callable
StrategyFactory = Any  # Callable[[EventBus, ExecutionClient], StrategyBase]


@dataclass(frozen=True)
class BacktestResult:
    """Complete result of a single backtest run."""

    metrics: PerformanceMetrics
    fills: list[Fill]
    equity_curve: list[tuple[Timestamp, Usd]]


class SimulatedExecutionClient:
    """Simulated execution client for backtesting.

    Implements the ExecutionClient Protocol so strategies can use the same
    interface in both live and backtest modes. Fills are simulated against
    candle data via the SimulatedBroker.
    """

    def __init__(
        self,
        broker: SimulatedBroker,
        event_bus: EventBus,
    ) -> None:
        self._broker = broker
        self._event_bus = event_bus
        self._open_orders: dict[Cloid, Order] = {}
        self._cloid_counter = 0

    async def submit_order(self, intent: OrderIntent) -> Order:
        """Submit an order intent. Returns the created Order.

        In backtest mode, the order is created but NOT immediately filled.
        The engine will attempt fills on subsequent candles.
        """
        cloid = intent.cloid or self._generate_cloid()
        order = Order(
            cloid=cloid,
            symbol=intent.symbol,
            side=intent.side,
            size=intent.size,
            price=intent.price,
            order_type=intent.order_type,
            time_in_force=intent.time_in_force,
            status=OrderStatus.SUBMITTED,
            strategy_id=intent.strategy_id,
            reduce_only=intent.reduce_only,
            created_at=datetime.now(UTC),
        )
        self._open_orders[cloid] = order

        self._event_bus.publish_sync(Event(event_type=EVENT_ORDER_SUBMITTED, payload=order, correlation_id=str(cloid)))
        return order

    async def cancel_order(self, cloid: str) -> bool:
        """Cancel an order by cloid."""
        c = Cloid(cloid)
        if c in self._open_orders:
            order = self._open_orders.pop(c)
            order.status = OrderStatus.CANCELLED
            self._event_bus.publish_sync(Event(event_type=EVENT_ORDER_CANCELLED, payload=order, correlation_id=cloid))
            return True
        return False

    async def cancel_all_orders(self, symbol: str | None = None) -> int:
        """Cancel all open orders, optionally filtered by symbol."""
        to_cancel = []
        for cloid, order in self._open_orders.items():
            if symbol is None or str(order.symbol) == symbol:
                to_cancel.append(cloid)

        for cloid in to_cancel:
            await self.cancel_order(str(cloid))
        return len(to_cancel)

    async def get_order(self, cloid: str) -> Order | None:
        return self._open_orders.get(Cloid(cloid))

    async def get_open_orders(self, symbol: str | None = None) -> list[Order]:
        if symbol is None:
            return list(self._open_orders.values())
        return [o for o in self._open_orders.values() if str(o.symbol) == symbol]

    def try_fill_orders(self, candle: Candle, fills_collector: list[Fill]) -> None:
        """Attempt to fill open orders against the given candle.

        Called by the engine after publishing each candle event.
        Successfully filled orders are removed from _open_orders and their
        Fill records are appended to fills_collector.
        """
        filled_cloids: list[Cloid] = []

        for cloid, order in self._open_orders.items():
            # Build an OrderIntent-like object for the broker
            intent = OrderIntent(
                symbol=order.symbol,
                side=order.side,
                size=order.remaining_size,
                price=order.price,
                order_type=order.order_type,
                time_in_force=order.time_in_force,
                strategy_id=order.strategy_id,
                reduce_only=order.reduce_only,
                cloid=cloid,
            )
            fill = self._broker.simulate_fill(intent, candle, cloid)
            if fill is not None:
                # Update order state
                order.status = OrderStatus.FILLED
                order.filled_size = order.size
                order.avg_fill_price = fill.price
                order.filled_at = datetime.now(UTC)
                filled_cloids.append(cloid)

                fills_collector.append(fill)
                self._event_bus.publish_sync(
                    Event(event_type=EVENT_ORDER_FILLED, payload=fill, correlation_id=str(cloid))
                )

        for cloid in filled_cloids:
            del self._open_orders[cloid]

    def _generate_cloid(self) -> Cloid:
        self._cloid_counter += 1
        return Cloid(f"bt_cloid_{self._cloid_counter}")


class BacktestEngine:
    """Main backtest orchestrator.

    Wires DataFeed, SimulatedBroker, SimulatedExecutionClient, and a strategy
    together. Replays historical data through the EventBus, lets the strategy
    react via on_event(), simulates fills, and computes performance metrics.
    """

    def __init__(
        self,
        fee_config: FeeConfig | None = None,
        slippage_config: SlippageConfig | None = None,
    ) -> None:
        self._fee_config = fee_config or FeeConfig()
        self._slippage_config = slippage_config or SlippageConfig()

    async def run(
        self,
        candles: list[Candle],
        funding_rates: list[FundingRate] | None,
        strategy_factory: StrategyFactory,
        initial_capital: Usd | None = None,
        slippage_mode: SlippageMode = SlippageMode.PESSIMISTIC,
    ) -> BacktestResult:
        """Run a backtest.

        Args:
            candles: Historical candle data (sorted by timestamp).
            funding_rates: Historical funding rate data (optional).
            strategy_factory: Callable that creates a strategy given
                (event_bus, execution_client). Example:
                lambda eb, ec: MyStrategy(StrategyId("test"), eb, ec)
            initial_capital: Starting equity in USDC.
            slippage_mode: Optimistic or pessimistic fill simulation.

        Returns:
            BacktestResult with metrics, fills, and equity curve.
        """
        if initial_capital is None:
            initial_capital = Usd(10_000.0)

        # Create isolated components for this run
        event_bus = EventBus(queue_maxsize=10_000)
        broker = SimulatedBroker(
            fee_config=self._fee_config,
            slippage_config=self._slippage_config,
            mode=slippage_mode,
        )
        execution_client = SimulatedExecutionClient(broker, event_bus)
        data_feed = DataFeed(candles, funding_rates, event_bus)

        # Portfolio state tracking
        equity = initial_capital
        peak_equity = initial_capital
        positions: dict[str, Position] = {}  # symbol -> Position
        all_fills: list[Fill] = []
        equity_curve: list[tuple[Timestamp, Usd]] = []
        total_funding = Usd(0.0)
        applied_funding: set[tuple[str, int]] = set()
        realized_trade_pnls: list[Usd] = []

        # Create and start strategy
        strategy = strategy_factory(event_bus, execution_client)
        await strategy.on_start()

        logger.info(
            "backtest_started",
            candles=data_feed.total_candles,
            initial_capital=float(initial_capital),
            slippage_mode=str(slippage_mode),
        )

        # Main simulation loop
        while data_feed.has_next:
            candle = data_feed.next_candle()
            if candle is None:
                break

            # 1. Fill orders submitted on earlier candles. This prevents a
            # strategy from seeing a candle's close/high/low and then filling
            # a newly-created order against that same candle.
            prev_fill_count = len(all_fills)
            execution_client.try_fill_orders(candle, all_fills)

            # 2. Apply fills to cash and positions.
            for fill in all_fills[prev_fill_count:]:
                equity = Usd(equity - fill.fee)
                realized = self._update_position(positions, fill)
                equity = Usd(equity + realized)
                if realized != Usd(0.0):
                    realized_trade_pnls.append(realized)

            # 3. Mark positions to the current close before recording equity.
            position = positions.get(str(candle.symbol))
            if position is not None:
                position.mark_price = candle.close

            # 4. Apply each exchange funding record at most once.
            if funding_rates:
                funding_amount = self._apply_funding_once(positions, funding_rates, candle.timestamp, applied_funding)
                if funding_amount != Usd(0.0):
                    equity = Usd(equity - funding_amount)
                    total_funding = Usd(total_funding + funding_amount)

            marked_equity = Usd(equity + self._unrealized_pnl(positions))
            if marked_equity > peak_equity:
                peak_equity = marked_equity
            equity_curve.append((candle.timestamp, marked_equity))

            # 5. Let the strategy react after this candle has been fully
            # accounted for. Orders created here are eligible next candle.
            candle_event = Event(
                event_type=EVENT_CANDLE_UPDATE,
                payload=candle,
                correlation_id=str(candle.symbol),
            )
            try:
                await strategy.on_event(candle_event)
            except Exception:
                logger.exception(
                    "backtest_strategy_error",
                    candle_ts=candle.timestamp,
                    symbol=str(candle.symbol),
                )

        # Stop strategy
        await strategy.on_stop()

        # Orders created by on_stop cannot be honestly filled without a later
        # market observation. Leave them open rather than introducing a
        # synthetic same-bar fill; final equity includes marked open positions.

        # Calculate metrics
        calculator = MetricsCalculator(
            fills=all_fills,
            equity_curve=equity_curve,
            initial_capital=initial_capital,
            funding_total=total_funding,
            trade_pnls=realized_trade_pnls,
        )
        metrics = calculator.calculate()

        logger.info(
            "backtest_complete",
            trades=metrics.trade_count,
            total_return=f"{metrics.total_return_pct:.4%}",
            max_drawdown=f"{metrics.max_drawdown_pct:.4%}",
            sharpe=f"{metrics.sharpe_ratio:.2f}",
        )

        return BacktestResult(
            metrics=metrics,
            fills=all_fills,
            equity_curve=equity_curve,
        )

    @staticmethod
    def _apply_fill_to_equity(fill: Fill, equity: Usd) -> Usd:
        """Update equity based on a fill's fee impact.

        Fees reduce equity (taker), rebates increase equity (maker).
        The position's PnL is tracked separately via mark-to-market.
        """
        return Usd(equity - fill.fee)

    @staticmethod
    def _update_position(positions: dict[str, Position], fill: Fill) -> Usd:
        """Update an average-cost position and return realized PnL."""
        key = str(fill.symbol)
        pos = positions.get(key)

        if pos is None:
            # New position
            signed_size = fill.size if fill.side == Side.BUY else -fill.size
            positions[key] = Position(
                symbol=fill.symbol,
                size=Size(signed_size),
                entry_price=fill.price,
                mark_price=fill.price,
            )
            return Usd(0.0)
        else:
            old_size = float(pos.size)
            signed_fill = float(fill.size) if fill.side == Side.BUY else -float(fill.size)
            new_size = old_size + signed_fill
            entry = float(pos.entry_price or fill.price)

            # Same direction: increase the position using VWAP.
            if old_size * signed_fill > 0:
                new_entry = (abs(old_size) * entry + abs(signed_fill) * float(fill.price)) / abs(new_size)
                pos.size = Size(new_size)
                pos.entry_price = Price(new_entry)
                pos.mark_price = fill.price
                return Usd(0.0)

            # Opposite direction: close all or part of the old position.
            closing_size = min(abs(old_size), abs(signed_fill))
            direction = 1.0 if old_size > 0 else -1.0
            realized = Usd((float(fill.price) - entry) * closing_size * direction)

            if abs(new_size) < 1e-12:
                del positions[key]
            elif old_size * new_size > 0:
                # Partial reduction: entry price must remain unchanged.
                pos.size = Size(new_size)
                pos.mark_price = fill.price
            else:
                # Position flipped: the residual starts at the flip fill.
                pos.size = Size(new_size)
                pos.entry_price = fill.price
                pos.mark_price = fill.price
            return realized

    @staticmethod
    def _apply_funding_once(
        positions: dict[str, Position],
        funding_rates: list[FundingRate],
        current_ts: int,
        applied: set[tuple[str, int]],
    ) -> Usd:
        """Apply due funding records exactly once and return account cost."""
        total = Usd(0.0)
        for pos in positions.values():
            if pos.is_flat or pos.mark_price is None:
                continue
            for rate in funding_rates:
                key = (str(rate.symbol), int(rate.timestamp))
                if key not in applied and rate.timestamp <= current_ts and str(rate.symbol) == str(pos.symbol):
                    funding = SimulatedBroker.apply_hourly_funding(pos, rate.funding_rate, pos.mark_price)
                    total = Usd(total + funding)
                    applied.add(key)
        return total

    @staticmethod
    def _unrealized_pnl(positions: dict[str, Position]) -> Usd:
        total = 0.0
        for position in positions.values():
            if position.entry_price is None or position.mark_price is None:
                continue
            total += float(position.size) * (float(position.mark_price) - float(position.entry_price))
        return Usd(total)

File: crates/trading/src/backtest/engine.rs
//! Backtest engine — orchestrates strategy, broker, and data feed, port of
//! `src/hypeedge/backtest/engine.py`.

use std::collections::HashMap;
use std::sync::Arc;

use hypeedge_domain::decimal::{Decimal, Price, Size, Usd};
use hypeedge_domain::enums::{OrderStatus, Side};
use hypeedge_domain::events::{DomainEvent, Event};
use hypeedge_domain::models::{Candle, Fill, FundingRate, Order, OrderIntent, Position};
use hypeedge_domain::traits::ExecutionClient;
use hypeedge_infra::event_bus::EventBus;

use super::broker::SimulatedBroker;
use super::metrics::MetricsCalculator;

/// Complete result of a single backtest run.
#[derive(Debug, Clone)]
pub struct BacktestResult {
    pub metrics: super::metrics::PerformanceMetrics,
    pub fills: Vec<Fill>,
    pub equity_curve: Vec<(i64, Usd)>,
}

/// The simulated execution client (implements `ExecutionClient`).
pub struct SimulatedExecutionClient {
    broker: Arc<std::sync::Mutex<SimulatedBroker>>,
    bus: Arc<EventBus>,
    open_orders: tokio::sync::Mutex<HashMap<String, Order>>,
    cloid_counter: std::sync::atomic::AtomicU64,
}

impl SimulatedExecutionClient {
    pub fn new(broker: Arc<std::sync::Mutex<SimulatedBroker>>, bus: Arc<EventBus>) -> Self {
        Self {
            broker,
            bus,
            open_orders: tokio::sync::Mutex::new(HashMap::new()),
            cloid_counter: std::sync::atomic::AtomicU64::new(0),
        }
    }

    /// Attempt to fill open orders against the given candle. Filled orders are
    /// removed and appended to `fills_collector`.
    pub async fn try_fill_orders(&self, candle: &Candle, fills_collector: &mut Vec<Fill>) {
        let mut open = self.open_orders.lock().await;
        let filled_cloids: Vec<String> = {
            let mut broker = self.broker.lock().unwrap();
            let mut filled = Vec::new();
            for (cloid, order) in open.iter_mut() {
                let intent = OrderIntent {
                    symbol: order.symbol.clone(),
                    side: order.side,
                    size: order.remaining_size(),
                    price: order.price,
                    order_type: order.order_type,
                    time_in_force: order.time_in_force,
                    strategy_id: order.strategy_id.clone(),
                    sub_account: None,
                    reduce_only: order.reduce_only,
                    cloid: Some(cloid.clone()),
                    client_id: None,
                    is_spot: false,
                    risk_reducing: false,
                    max_slippage_bps: 50,
                };
                if let Some(fill) = broker.simulate_fill(&intent, candle, cloid) {
                    order.status = OrderStatus::Filled;
                    order.filled_size = order.size;
                    order.avg_fill_price = Some(fill.price);
                    order.filled_at = Some(chrono::Utc::now());
                    filled.push(cloid.clone());
                    fills_collector.push(fill.clone());
                    let _ = self.bus.publish_sync(Arc::new(
                        Event::new(DomainEvent::OrderFilled(order.clone()))
                            .with_correlation_id(cloid.clone()),
                    ));
                }
            }
            filled
        };
        for cloid in filled_cloids {
            open.remove(&cloid);
        }
    }

    fn generate_cloid(&self) -> String {
        let n = self
            .cloid_counter
            .fetch_add(1, std::sync::atomic::Ordering::SeqCst)
            + 1;
        format!("bt_cloid_{n}")
    }
}

#[async_trait::async_trait]
impl ExecutionClient for SimulatedExecutionClient {
    async fn submit_order(
        &self,
        intent: OrderIntent,
        _deferred: Option<bool>,
    ) -> Result<Order, hypeedge_domain::error::HypeEdgeError> {
        let cloid = intent
            .cloid
            .clone()
            .unwrap_or_else(|| self.generate_cloid());
        let order = Order {
            cloid: cloid.clone(),
            symbol: intent.symbol.clone(),
            side: intent.side,
            size: intent.size,
            price: intent.price,
            order_type: intent.order_type,
            time_in_force: intent.time_in_force,
            status: OrderStatus::Submitted,
            strategy_id: intent.strategy_id.clone(),
            sub_account: None,
            reduce_only: intent.reduce_only,
            is_spot: false,
            risk_reducing: false,
            max_slippage_bps: 50,
            exchange_oid: None,
            filled_size: Size::ZERO,
            avg_fill_price: None,
            submitted_at: None,
            acknowledged_at: None,
            filled_at: None,
            error_message: None,
            created_at: chrono::Utc::now(),
        };
        self.open_orders
            .lock()
            .await
            .insert(cloid.clone(), order.clone());
        let _ = self.bus.publish_sync(Arc::new(
            Event::new(DomainEvent::OrderSubmitted(order.clone())).with_correlation_id(cloid),
        ));
        Ok(order)
    }

    async fn cancel_order(
        &self,
        cloid: &str,
    ) -> Result<bool, hypeedge_domain::error::HypeEdgeError> {
        let mut open = self.open_orders.lock().await;
        if let Some(mut order) = open.remove(cloid) {
            order.status = OrderStatus::Cancelled;
            let _ = self.bus.publish_sync(Arc::new(
                Event::new(DomainEvent::OrderCancelled(order.clone()))
                    .with_correlation_id(cloid.to_string()),
            ));
            Ok(true)
        } else {
            Ok(false)
        }
    }

    async fn cancel_all_orders(
        &self,
        symbol: Option<&str>,
    ) -> Result<u64, hypeedge_domain::error::HypeEdgeError> {
        let mut open = self.open_orders.lock().await;
        let keys: Vec<String> = open
            .iter()
            .filter(|(_, o)| symbol.is_none() || o.symbol == symbol.unwrap())
            .map(|(k, _)| k.clone())
            .collect();
        let count = keys.len() as u64;
        for k in keys {
            open.remove(&k);
        }
        Ok(count)
    }

    async fn get_order(
        &self,
        cloid: &str,
    ) -> Result<Option<Order>, hypeedge_domain::error::HypeEdgeError> {
        Ok(self.open_orders.lock().await.get(cloid).cloned())
    }

    async fn get_open_orders(
        &self,
        symbol: Option<&str>,
    ) -> Result<Vec<Order>, hypeedge_domain::error::HypeEdgeError> {
        let open = self.open_orders.lock().await;
        Ok(open
            .values()
            .filter(|o| symbol.is_none() || o.symbol == symbol.unwrap())
            .cloned()
            .collect())
    }
}

/// The main backtest orchestrator. Fee/slippage live on the caller-built
/// [`SimulatedBroker`]; the engine only replays candles, fills orders from the
/// injected [`SimulatedExecutionClient`], and computes metrics (A24).
pub struct BacktestEngine;

impl Default for BacktestEngine {
    fn default() -> Self {
        Self::new()
    }
}

impl BacktestEngine {
    pub fn new() -> Self {
        BacktestEngine
    }

    #[allow(clippy::too_many_arguments)]
    pub async fn run(
        &self,
        candles: Vec<Candle>,
        funding_rates: Option<Vec<FundingRate>>,
        execution: Arc<SimulatedExecutionClient>,
        initial_capital: Usd,
    ) -> BacktestResult {
        let mut equity = initial_capital;
        let mut peak_equity = initial_capital;
        let mut positions: HashMap<String, Position> = HashMap::new();
        let mut all_fills: Vec<Fill> = Vec::new();
        let mut equity_curve: Vec<(i64, Usd)> = Vec::new();
        let mut total_funding = Usd::ZERO;
        let mut applied_funding: std::collections::HashSet<(String, i64)> = Default::default();
        let mut realized_trade_pnls: Vec<Usd> = Vec::new();
        let mut total_fees = Usd::ZERO;

        // A24: fill orders from the caller-supplied simulated client — the same
        // one the strategy submits through. The pre-fix code built a private
        // client nothing ever submitted to, so run() always returned zero fills.
        for candle in &candles {
            // 1. Fill orders submitted on earlier candles.
            let prev_fill_count = all_fills.len();
            execution.try_fill_orders(candle, &mut all_fills).await;

            // 2. Apply fills to cash and positions.
            for fill in &all_fills[prev_fill_count..] {
                equity = Usd::new(equity.inner() - fill.fee.inner());
                total_fees = Usd::new(total_fees.inner() + fill.fee.inner().abs());
                let realized = Self::update_position(&mut positions, fill);
                equity = Usd::new(equity.inner() + realized.inner());
                if !realized.is_zero() {
                    realized_trade_pnls.push(realized);
                }
            }

            // 3. Mark positions to the current close.
            if let Some(position) = positions.get_mut(&candle.symbol) {
                position.mark_price = Some(candle.close);
            }

            // 4. Apply each funding record at most once.
            if let Some(rates) = &funding_rates {
                let funding_amount = Self::apply_funding_once(
                    &positions,
                    rates,
                    candle.timestamp,
                    &mut applied_funding,
                );
                if !funding_amount.is_zero() {
                    equity = Usd::new(equity.inner() - funding_amount.inner());
                    total_funding = Usd::new(total_funding.inner() + funding_amount.inner());
                }
            }

            let marked_equity = Usd::new(equity.inner() + Self::unrealized_pnl(&positions).inner());
            if marked_equity.inner() > peak_equity.inner() {
                peak_equity = marked_equity;
            }
            equity_curve.push((candle.timestamp, marked_equity));
        }

        // Calculate metrics.
        let calculator = MetricsCalculator::new(
            equity_curve.clone(),
            initial_capital,
            total_funding,
            total_fees,
            realized_trade_pnls,
        );
        let metrics = calculator.calculate();

        BacktestResult {
            metrics,
            fills: all_fills,
            equity_curve,
        }
    }

    fn update_position(positions: &mut HashMap<String, Position>, fill: &Fill) -> Usd {
        let key = fill.symbol.clone();
        let signed_fill = if fill.side == Side::Buy {
            fill.size.inner()
        } else {
            -fill.size.inner()
        };
        match positions.get_mut(&key) {
            None => {
                positions.insert(
                    key,
                    Position {
                        symbol: fill.symbol.clone(),
                        size: Size::new(signed_fill),
                        entry_price: Some(fill.price),
                        mark_price: Some(fill.price),
                        unrealized_pnl: None,
                        leverage: 1,
                        liquidation_price: None,
                        sub_account: None,
                        strategy_id: None,
                    },
                );
                Usd::ZERO
            }
            Some(pos) => {
                let old_size = pos.size.inner();
                let new_size = old_size + signed_fill;
                let entry = pos
                    .entry_price
                    .map(|p| p.inner())
                    .unwrap_or(fill.price.inner());

                // Same direction: increase using VWAP.
                if (old_size.is_positive() && signed_fill.is_positive())
                    || (old_size.is_negative() && signed_fill.is_negative())
                {
                    let new_entry = (old_size.abs() * entry
                        + signed_fill.abs() * fill.price.inner())
                        / new_size.abs();
                    pos.size = Size::new(new_size);
                    pos.entry_price = Some(Price::new(new_entry));
                    pos.mark_price = Some(fill.price);
                    return Usd::ZERO;
                }

                // Opposite direction: close all or part.
                let closing_size = old_size.abs().min(signed_fill.abs());
                let direction = if old_size > Decimal::ZERO {
                    Decimal::ONE
                } else {
                    -Decimal::ONE
                };
                let realized = (fill.price.inner() - entry) * closing_size * direction;

                if new_size.abs() < Decimal::from_str_lenient("0.000000000001").unwrap() {
                    positions.remove(&key);
                } else if old_size.is_positive() == new_size.is_positive() {
                    // Partial reduction: entry unchanged.
                    pos.size = Size::new(new_size);
                    pos.mark_price = Some(fill.price);
                } else {
                    // Flip: residual starts at the flip fill.
                    pos.size = Size::new(new_size);
                    pos.entry_price = Some(fill.price);
                    pos.mark_price = Some(fill.price);
                }
                Usd::new(realized)
            }
        }
    }

    fn apply_funding_once(
        positions: &HashMap<String, Position>,
        funding_rates: &[FundingRate],
        current_ts: i64,
        applied: &mut std::collections::HashSet<(String, i64)>,
    ) -> Usd {
        let mut total = Usd::ZERO;
        for position in positions.values() {
            if position.is_flat() || position.mark_price.is_none() {
                continue;
            }
            for rate in funding_rates {
                let key = (rate.symbol.clone(), rate.timestamp);
                if !applied.contains(&key)
                    && rate.timestamp <= current_ts
                    && rate.symbol == position.symbol
                {
                    let funding = SimulatedBroker::apply_hourly_funding(
                        position,
                        rate.funding_rate,
                        position.mark_price.unwrap(),
                    );
                    total = Usd::new(total.inner() + funding.inner());
                    applied.insert(key);
                }
            }
        }
        total
    }

    fn unrealized_pnl(positions: &HashMap<String, Position>) -> Usd {
        let mut total = Decimal::ZERO;
        for position in positions.values() {
            let (Some(entry), Some(mark)) = (position.entry_price, position.mark_price) else {
                continue;
            };
            total += position.size.inner() * (mark.inner() - entry.inner());
        }
        Usd::new(total)
    }
}
